### PR TITLE
gotohelm: many many fixes for numeric handling

### PR DESCRIPTION
**Pull request by chrisseto**
#### 2b401bb0448a0087c0513f16496cea3c46c65f90 gotohelm: fix float constants

Floating point constants in go, when possible, are represented as division of
rational numbers. For example, 0.1 would be represented as 1/10. This syntax is
passed directly to helm which does not support such division.

This commit handles floating point constants and forcibly represents them as
decimals in all cases.

#### 3f62f3a40be426d91265c6d816b97a7df9c38883 gotohelm: automatically cast when possible

Prior to this commit, writing gotohelm code need to be very tactical when
dealing with numeric values. This lead to type casts being littered around the
codebase and likely a handful of unnoticed bugs.

This commit upgrades the transpiler to inject type casts into transpiled code
at function boundaries and struct field references.

While this solution is not perfect, it should dramatically reduce needless type
casting and make code slightly more idiomatic.

#### cd0991028893359424fcf74c589bbd6e32884f60 gotohelm: add numeric helper functions

This commit adds the `AsNumeric` and `AsIntegral` helper functions to the
`helmette` package to aid working with numeric values in gotohelm code.

A future commit will disallow type assertions and type checks of numeric values
in favor of these helpers as they are the closest approximation to keeping helm
and go behavior the same.

#### f4762696d204621d09f53aaffde4366e5a39347c gotohelm: Ban numeric type casting

Prior to this commit is was value to use type checks or assertions within
gotohelm code with numeric values. For example: `x.(int)`. Due to the numerous
limitations of go templates and helm, it's not feasible to fully support such
expressions as there's no type information that could be used to distinguish
between ints, int64s, and float64s.

This commit causes the transpiler to fail if it encounters such a statement and
instead points users to the recently added `AsNumeric` and `AsIntegral`
functions.

### CODE REVIEW

**Comment by chrisseto** [comment]
Happy to split these up into smaller PRs but I figured I'd bundle them together so it's easier to see the "big picture".

File: pkg/gotohelm/testdata/src/example/sprig/shims.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	inputs = append(inputs, int(10), 1.5, dot.Values["numeric"])
+
+	outputs := []any{}
+	for _, in := range inputs {
+		value, isIntegral := helmette.AsIntegral[int](in)
+
+		outputs = append(outputs, []any{in, value, isIntegral})
+	}
+
+	return nil
```
Comment (RafalKorepta) [inline pkg/gotohelm/testdata/src/example/sprig/shims.go:47]:
```suggestion
	return outputs
```

File: pkg/gotohelm/transpiler.go [hunk 2]
```diff
@@ -1033,6 +1024,28 @@ func (t *Transpiler) transpileTypeRepr(typ types.Type) Node {
...
+	// them. For now, it's easier to turn constants into their
+	// definitions.
+
+	if c.Val().Kind() != constant.Float {
+		return &Literal{Value: c.Val().ExactString()}
+	}
+
+	// Floats need a bit of special handling as they _might_ be represented as
+	// a fractions which we don't want.
+	if asFloat, ok := constant.Float64Val(c.Val()); ok || asFloat > 0.00001 {
```
[the remark this answers is not in the document]
Comment (chrisseto) [reply pkg/gotohelm/transpiler.go:1063]: Ah, yeah I should add a note! It's a completely arbitrary value for permitting numbers that don't return true from `Float64Val` but should still be expressible. I think I hit a few of those but I'm struggling to recall. I'll take it out and see what happens and write a novel about what it is once I figure it out again.

File: pkg/gotohelm/testdata/src/example/sprig/shims.go (hunk 1 above)
Comment (chrisseto) [reply pkg/gotohelm/testdata/src/example/sprig/shims.go]: Fixed!

File: pkg/gotohelm/transpiler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisseto) [reply pkg/gotohelm/transpiler.go]: I ended up removing this condition as there's not a good way to handle it IMO. I added a _very_ long comment about exactly that 😓

**Review by RafalKorepta — APPROVED** [review]
LGTM

**Pull request merged**